backend/app/analysis/find_vanishing_point.py

```diff
@@ -0,0 +1,227 @@
+intersect, which should be the vanishing point
+"""
+
+import numpy as np
+import cv2
+
+from app.models import Photo
+
+MODEL = Photo
+
+
+def analyze(photo: Photo):
+    """
+    Given a photo, returns the coordinate of the vanishing point,
+    where the vanishing point is the point that has the minimum sum of
+    distances to all detected lines in the photo
+    """
+    image = photo.get_image_data()
+    # Convert image to grayscale
+    # (Changes image array shape from (height, width, 3) to (height, width))
+    # (Pixels (image[h][w]) will be a value from 0 to 255)
+    grayscale_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+    lines = auto_canny(grayscale_image)
+
+    filter_lines = []
+    # Filters out vertical and horizontal lines
+    for line in lines[1]:
+        try:
+            point_1_x, point_1_y, point_2_x, point_2_y = line[0]
+
+            # find angle of line from horizontal
+            theta = abs(np.arctan((point_2_y - point_1_y) / (point_2_x - point_1_x)))
+
+            # only put line in filter_lines if the angle is NOT within epsilon of
+            # horizontal/vertical
+            epsilon = np.pi / 16
+            if (
+                (point_2_x - point_1_x) != 0
+                and abs(theta - np.pi / 2) > 2 * epsilon
+                and theta > epsilon
+            ):
+                filter_lines.append({
+                    '1_x': int(point_1_x),
+                    '1_y': int(point_1_y),
+                    '2_x': int(point_2_x),
+                    '2_y': int(point_2_y),
+                })
+
+        except ZeroDivisionError:
+            pass
+
+    van_point = find_van_coord_intersections(filter_lines)
+
+    return {
+        'vanishing_point_coord': van_point,
+        'line_coords': filter_lines,
+    }
+
+def auto_canny(image):
+    """
+    Applies the Canny and HoughLinesP functions from OpenCV to the given
+    image and returns the result
+
+    :param image: an image
+    :return: a list containing the binary image from Canny and the set of
+             lines from HoughLinesP
+
+    TODO: Optimize parameters for Canny and HoughLinesP instead of hard-coding them
+    """
+
+    edged = cv2.Canny(image, 150, 255, 3)
+    lines = cv2.HoughLinesP(edged, 1, np.pi / 180, 80, 30, maxLineGap=100)
+
+    # return the edges and lines
+    return [edged, lines]
+
+
+def find_van_coord_intersections(lines):
+    """
+    Find the approximate vanishing point of the image by choosing the most representative point
+    from the largest cluster of intersections between lines in 'lines'
+    :param lines: list of lists containing start and end points of all filtered lines
+    :return: the average coordinate of the largest cluster of intersections
+    """
+    intersections = {}
+    tolerance = 30
+    # print(len(lines))
+    # return None if the image likely has many unnecessary lines (e.g. if there's a tree)
+    if len(lines) > 300 or len(lines) < 2:
```

Comment:
please name all magic numbers. (300 is an arbitrary bound, so if you call it something like MAX_ALLOWED_LINES, then readers will know why it's there.)